### PR TITLE
[Unix] Fix the OpenGL undetected on Linux, caused by GTK3+ detection snafu

### PR DESCRIPTION
**Pull request by ebachard**
### Pull Request Readiness Checklist

OpenCV version : master (4.5.0) 
OS : Linux Intel, x86_64  (Mint 19.2)
Configure file + option  is attached 
Compiler : g++ 7.x

I didn"t check the history, but looks like somebody truncated the HAVE_GTK3 into GTK3 

Hello, this little change should fix the never detected OpenGL on Linux, caused by a macro name collision. 

Feel free to integrate and backport (or not) my change, working on both master and 3.4.12 version, Linux x86_64.

Last but not least, this fix is very probably linked to #18170 

Thanks. 
Eric Bachard

The log including the change :
````
-- General configuration for OpenCV 4.5.0-dev =====================================
--   Version control:               4.5.0-91-g7459613623-dirty
-- 
--   Platform:
--     Timestamp:                   2020-11-05T22:36:45Z
--     Host:                        Linux 4.15.23 x86_64
--     CMake:                       3.15.3
--     CMake generator:             Unix Makefiles
--     CMake build tool:            /usr/bin/make
--     Configuration:               Release
-- 
--   CPU/HW features:
--     Baseline:                    SSE SSE2 SSE3
--       requested:                 SSE3
--     Dispatched code generation:  SSE4_1 SSE4_2 FP16 AVX AVX2 AVX512_SKX
--       requested:                 SSE4_1 SSE4_2 AVX FP16 AVX2 AVX512_SKX
--       SSE4_1 (17 files):         + SSSE3 SSE4_1
--       SSE4_2 (2 files):          + SSSE3 SSE4_1 POPCNT SSE4_2
--       FP16 (1 files):            + SSSE3 SSE4_1 POPCNT SSE4_2 FP16 AVX
--       AVX (5 files):             + SSSE3 SSE4_1 POPCNT SSE4_2 AVX
--       AVX2 (31 files):           + SSSE3 SSE4_1 POPCNT SSE4_2 FP16 FMA3 AVX AVX2
--       AVX512_SKX (7 files):      + SSSE3 SSE4_1 POPCNT SSE4_2 FP16 FMA3 AVX AVX2 AVX_512F AVX512_COMMON AVX512_SKX
-- 
--   C/C++:
--     Built as dynamic libs?:      YES
--     C++ standard:                11
--     C++ Compiler:                /usr/bin/g++-7  (ver 7.5.0)
--     C++ flags (Release):         -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Winit-self -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden -fvisibility-inlines-hidden -O3 -DNDEBUG  -DNDEBUG
--     C++ flags (Debug):           -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Winit-self -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden -fvisibility-inlines-hidden -g  -O0 -DDEBUG -D_DEBUG
--     C Compiler:                  /usr/bin/gcc-7
--     C flags (Release):           -fsigned-char -W -Wall -Werror=return-type -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wuninitialized -Winit-self -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden -O3 -DNDEBUG  -DNDEBUG
--     C flags (Debug):             -fsigned-char -W -Wall -Werror=return-type -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wuninitialized -Winit-self -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden -g  -O0 -DDEBUG -D_DEBUG
--     Linker flags (Release):      -Wl,--exclude-libs,libippicv.a -Wl,--exclude-libs,libippiw.a   -Wl,--gc-sections -Wl,--as-needed  
--     Linker flags (Debug):        -Wl,--exclude-libs,libippicv.a -Wl,--exclude-libs,libippiw.a   -Wl,--gc-sections -Wl,--as-needed  
--     ccache:                      YES
--     Precompiled headers:         NO
--     Extra dependencies:          dl m pthread rt /usr/lib/x86_64-linux-gnu/libGL.so /usr/lib/x86_64-linux-gnu/libGLU.so
--     3rdparty dependencies:
-- 
--   OpenCV modules:
--     To be built:                 calib3d core dnn features2d flann gapi highgui imgcodecs imgproc java ml objdetect photo python2 python3 stitching ts video videoio
--     Disabled:                    world
--     Disabled by dependency:      -
--     Unavailable:                 js
--     Applications:                tests perf_tests examples apps
--     Documentation:               NO
--     Non-free algorithms:         NO
-- 
--   GUI: 
--     GTK+:                        YES (ver 2.24.32)
--       GThread :                  YES (ver 2.56.4)
--       GtkGlExt:                  YES (ver 1.2.0)
--     OpenGL support:              YES (/usr/lib/x86_64-linux-gnu/libGL.so /usr/lib/x86_64-linux-gnu/libGLU.so)
--     VTK support:                 NO
-- 

(and so on ... )
````

Before it was :
````
-- General configuration for OpenCV 4.5.0-dev =====================================
--   Version control:               4.5.0-91-g7459613623
-- 
--   Platform:
--     Timestamp:                   2020-11-05T22:35:24Z
--     Host:                        Linux 4.15.23 x86_64
--     CMake:                       3.15.3
--     CMake generator:             Unix Makefiles
--     CMake build tool:            /usr/bin/make
--     Configuration:               Release
-- 
--   CPU/HW features:
--     Baseline:                    SSE SSE2 SSE3
--       requested:                 SSE3
--     Dispatched code generation:  SSE4_1 SSE4_2 FP16 AVX AVX2 AVX512_SKX
--       requested:                 SSE4_1 SSE4_2 AVX FP16 AVX2 AVX512_SKX
--       SSE4_1 (17 files):         + SSSE3 SSE4_1
--       SSE4_2 (2 files):          + SSSE3 SSE4_1 POPCNT SSE4_2
--       FP16 (1 files):            + SSSE3 SSE4_1 POPCNT SSE4_2 FP16 AVX
--       AVX (5 files):             + SSSE3 SSE4_1 POPCNT SSE4_2 AVX
--       AVX2 (31 files):           + SSSE3 SSE4_1 POPCNT SSE4_2 FP16 FMA3 AVX AVX2
--       AVX512_SKX (7 files):      + SSSE3 SSE4_1 POPCNT SSE4_2 FP16 FMA3 AVX AVX2 AVX_512F AVX512_COMMON AVX512_SKX
-- 
--   C/C++:
--     Built as dynamic libs?:      YES
--     C++ standard:                11
--     C++ Compiler:                /usr/bin/g++-7  (ver 7.5.0)
--     C++ flags (Release):         -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Winit-self -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden -fvisibility-inlines-hidden -O3 -DNDEBUG  -DNDEBUG
--     C++ flags (Debug):           -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Winit-self -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden -fvisibility-inlines-hidden -g  -O0 -DDEBUG -D_DEBUG
--     C Compiler:                  /usr/bin/gcc-7
--     C flags (Release):           -fsigned-char -W -Wall -Werror=return-type -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wuninitialized -Winit-self -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden -O3 -DNDEBUG  -DNDEBUG
--     C flags (Debug):             -fsigned-char -W -Wall -Werror=return-type -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wuninitialized -Winit-self -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden -g  -O0 -DDEBUG -D_DEBUG
--     Linker flags (Release):      -Wl,--exclude-libs,libippicv.a -Wl,--exclude-libs,libippiw.a   -Wl,--gc-sections -Wl,--as-needed  
--     Linker flags (Debug):        -Wl,--exclude-libs,libippicv.a -Wl,--exclude-libs,libippiw.a   -Wl,--gc-sections -Wl,--as-needed  
--     ccache:                      YES
--     Precompiled headers:         NO
--     Extra dependencies:          dl m pthread rt
--     3rdparty dependencies:
-- 
--   OpenCV modules:
--     To be built:                 calib3d core dnn features2d flann gapi highgui imgcodecs imgproc java ml objdetect photo python2 python3 stitching ts video videoio
--     Disabled:                    world
--     Disabled by dependency:      -
--     Unavailable:                 js
--     Applications:                tests perf_tests examples apps
--     Documentation:               NO
--     Non-free algorithms:         NO
-- 
--   GUI: 
--     GTK+:                        YES (ver 3.22.30)
--       GThread :                  YES (ver 2.56.4)
--       GtkGlExt:                  NO
--     OpenGL support:              NO
--     VTK support:                 NO

````

Attached  : my configure script.

[configure_opencv.txt](https://github.com/opencv/opencv/files/5497857/configure_opencv.txt)

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ X] I agree to contribute to the project under Apache 2 License.
- [ X] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ X] The PR is proposed to proper branch 
- [ X] There is reference to original bug report and related work
- [ X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [X ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by ebachard** [comment]
Hello, 

Feel free to add or not this little change.

File: cmake/OpenCVFindLibsGUI.cmake [hunk 1]
```diff
@@ -38,7 +38,7 @@ endif()
 ocv_clear_vars(HAVE_GTK HAVE_GTK3 HAVE_GTHREAD HAVE_GTKGLEXT)
 if(WITH_GTK AND NOT HAVE_QT)
   if(NOT WITH_GTK_2_X)
-    ocv_check_modules(GTK3 gtk+-3.0)
+    ocv_check_modules(HAVE_GTK3 gtk+-3.0)
     if(HAVE_GTK3)
```
Comment (alalek) [inline cmake/OpenCVFindLibsGUI.cmake:42]: This change is wrong.
Check `ocv_check_modules()` implementation - it sets corresponding `HAVE_*` vars.

P.S. Always clean CMake cache (clean build directory) during dependencies change.

**Comment by alalek** [comment]
```
GTK+:                        YES (ver 2.24.32)
vs
GTK+:                        YES (ver 3.22.30)
```

You have break GTK3 detection.
Closing...

P.S. Look at `WITH_GTK_2_X` variable.

**Comment by ebachard** [comment]
Hello,

No : the guy who introduced GTK3 macro brokeOPENGL detection on Linux.  Verified with 4.5.0 and 3.4.12


And I didn't break GTK3 detection, since you cannot use OpenGL with gtkglext, since gtkglext needs gtk+2 


The problem with OpenCV, is that you modiffy the API inside the same major version (this should be of course avoided)

Worse, you regularly break OpenCV with unverified changes, and you never fix them correctly.
 

Last, If you look at the attached file, I am forced to uuse GTK2 with OpenGL ...

What a bad maintenance ...

````
        -DHAVE_GTK=ON \
        -DHAVE_GTKGLEXT=ON \
        -DHAVE_GTK3=NO \
        -DHAVE_GTK=YES ..
````

**Comment by mshabunin** [comment]
@ebachard , GTK3 + OpenGL is not supported yet (see https://github.com/opencv/opencv/commit/c1cd70e25abc1391aa3055fd09a22ae5197d7292), feel free to propose a patch if you think this is wrong.

Please note that `HAVE_*` variables are considered _internal_ and should not be modified by user in most cases. Configuration should be changed using _public_ `WITH_*` options and hints in form of environment variables and cmake options, e.g. `ABC_DIR` for `find_package(ABC)`.

**Comment by ebachard** [comment]
@mshabushin : I will have a look one day. 

If I'm not wrong, the main problem is to allow OpenGL + GTK3+w without gtkglext, because 3.x+ is common on Linux, and gtkglext seems to be obsolete. 

Waiting, using my change OpenGL and VTK work very well on both aarch64 and x86_64 Linux using 3.4.12 (broken with cv::CAP_GSTREAMER, not findind the webcam, but not only), and 3.4.0 (fast and working, but less secure).

To be continued

**Comment by ebachard** [comment]
FOR THE RECORD 

OS : Linux + OpenCV 3.4.x + OpenGL + GTK2  (no Qt in highgui)

Build successfully completed with last git checkout (2021 july 13th ) of the 3.4 branch. I'll start to have a look at the GTK3 change to make OpenGL work with gtk+3.x.  (please be patient)

FYI, the trick was to add :   -DHAVE_GTKGLEXT=YES 

Suggestion : instead of use HAVE_GTKGLEXT (defined nowhere IIRC), I'd suggest to add a WITH_GTKGLEXT configure option. Means add (e.g.)  -DWITH_GTKGLEXT=YES will set HAVE_GTKGLEXT, since this last macro should be private. The HAVE_GTKGLEXT macro appears in ${OPENCV_ROOT_DIR}/modules/highgui/src/window_gtk.cpp

Below, the details:

~/Devel/OpenCV/opencv $ git checkout 3.4
Basculement sur la branche '3.4'

~/Devel/OpenCV/opencv $ git status
Sur la branche 3.4

$ ./opencv_version
4.5.3-dev

Below, the cmake command line, to help people building 3.4 + OpenGL on Linux

 ~/Devel/OpenCV/opencv/build $  cat configure_opencv.txt 

#!/bin/bash

cmake   -DCMAKE_INSTALL_PREFIX=/usr/local \
        -DBUILD_SHARED_LIBS=YES \
        -DBUILD_EXAMPLES=ON \
        -DHAVE_VA_INTEL=YES \
        -DOPENCV_VERSION=Release \
        -DOPENCV_EXTRA_MODULES_PATH=../../opencv_contrib/modules/ \
        -DBUILD_NEW_PYTHON_SUPPORT=ON  \
        -DWITH_TBB=OFF \
        -DWITH_V4L=ON \
        -DWITH_OPENGL=TRUE \
        -DOpenGL_GL_PREFERENCE=GLVND \
        -DWITH_QT=OFF \
        -DHAVE_FFMPEG_RESAMPLE=YES \
        -DHAVE_GTKGLEXT=YES \
        -DWITH_GTK_2_X=YES \
        ..

**Comment by ebachard** [comment]
Last but not least, add a very helpfull link: https://developer.gnome.org/gtk3/stable/GtkGLArea.html

**Comment by ebachard** [comment]
TODO : fill a new issue, and RFC with a working gtk+3 implementation

**Pull request closed without merging**